### PR TITLE
harbor scale handler implements

### PR DESCRIPTION
**Pull request by wangcanfengxs**

### CODE REVIEW

File: controllers/harbor/harbor.go [hunk 1]
```diff
@@ -30,18 +30,28 @@ func (harbor *HarborReconciler) Reconcile() (*lcm.CRStatus, error) {
...
 		} else {
-			// TODO given clear reason and message.
-			return harborCRNotReadyStatus("", ""), err
+			return harborClusterCRUnknownStatus(), err
 		}
 	}
-	return nil, nil
-}
+	harbor.CurrentHarborCR = &harborCR
+	isScalingEvent := harbor.isScalingEvent(harbor.HarborCluster, &harborCR)
```
Comment (steven-zou) [inline controllers/harbor/harbor.go:37]: `isScalingEvent` and `isUpdatingEvent` are based on the similar parameters. So, maybe we can define a function like 

```go
 // return different reconcile const values
//  "Scaling"
//  "Udating"
  harbor.checkReconcileKind(harbor.HarborCluster, &harborCR)
```

Then here we can use `switch` to organize the actions (Scale()/Update())

File: controllers/harbor/harbor.go [hunk 2]
```diff
@@ -30,18 +30,28 @@ func (harbor *HarborReconciler) Reconcile() (*lcm.CRStatus, error) {
 		if errors.IsNotFound(err) {
 			return harbor.Provision()
 		} else {
-			// TODO given clear reason and message.
-			return harborCRNotReadyStatus("", ""), err
+			return harborClusterCRUnknownStatus(), err
```
Comment (steven-zou) [inline controllers/harbor/harbor.go:33]: Why the error here is `Unknown`?

File: controllers/harbor/scale.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	desiredReplicas := int32(desired.Spec.Replicas)
+
+	if current.Spec.Components.Core != nil && current.Spec.Components.Core.Replicas != nil {
+		coreReplicas := current.Spec.Components.Core.Replicas
+		if desiredReplicas != *coreReplicas {
+			return true
+		}
+	}
+
+	if current.Spec.Components.Portal != nil && current.Spec.Components.Portal.Replicas != nil {
```
Comment (steven-zou) [inline controllers/harbor/scale.go:66]: I remembered we discussed using a replica of harbor-core as the primary replica value, why do you need to check replicas of other components?

Comment (wangcanfengxs) [reply controllers/harbor/scale.go:66]: I think when we refactored to `Scale` method, we can check any component replicas.

Comment (steven-zou) [reply controllers/harbor/scale.go:66]: replica of harbor-core represents the overall replica of the harbor cluster except for jobservice, so I think only its replica can influence the overall replica settings.

File: controllers/harbor/harbor.go [hunk 4]
```diff
@@ -30,18 +36,79 @@ func (harbor *HarborReconciler) Reconcile() (*lcm.CRStatus, error) {
...
+	err = harbor.Get(harbor.getHarborCRNamespacedName(), &harborCR)
+	if err != nil {
+		return harborClusterCRUnknownStatus(), err
+	}
+	return harborClusterCRStatus(&harborCR), nil
 }
 
-func (harbor *HarborReconciler) Delete() (*lcm.CRStatus, error) {
-	panic("implement me")
+func unsetReplicas(harbor *v1alpha1.Harbor) {
```
Comment (steven-zou) [inline controllers/harbor/harbor.go:58]: can we build a component array (the elements of the array should implement the same interface) to do the replica setting logic? Code does not look like 'clean', what do you think?

File: controllers/harbor/harbor.go (hunk 2 above)
Comment (steven-zou) [reply controllers/harbor/harbor.go:39]: What's the reason?

File: controllers/harbor/harbor.go [hunk 5]
```diff
@@ -30,18 +36,82 @@ func (harbor *HarborReconciler) Reconcile() (*lcm.CRStatus, error) {
...
+	}
+	return harborClusterCRStatus(&harborCR), nil
 }
 
-func (harbor *HarborReconciler) Delete() (*lcm.CRStatus, error) {
-	panic("implement me")
+// unsetReplicas will set replicas to nil for all components in v1alpha1.Harbor.
+// This is a helper method to check whether harbor cr is equal expect replicas.
+func unsetReplicas(harbor *v1alpha1.Harbor) {
+	if harbor.Spec.Components.Core != nil {
```
Comment (steven-zou) [inline controllers/harbor/harbor.go:61]: The definitions of Harbor itself including:

```go
type HarborSpec struct {
  // ...
  Components HarborComponents `json:"components,omitempty"`
  // ...
}

type HarborComponents struct {
	// +optional
	Core *CoreComponent `json:"core,omitempty"`

	// +optional
	Portal *PortalComponent `json:"portal,omitempty"`

	// +optional
	Registry *RegistryComponent `json:"registry,omitempty"`

	// +optional
	JobService *JobServiceComponent `json:"jobService,omitempty"`

	// +optional
	ChartMuseum *ChartMuseumComponent `json:"chartMuseum,omitempty"`

	// +optional
	Clair *ClairComponent `json:"clair,omitempty"`

	// +optional
	Notary *NotaryComponent `json:"notary,omitempty"`
}

// Use core component as an example

type CoreComponent struct {
	HarborDeployment `json:",inline"`

	// +kubebuilder:validation:Required
	DatabaseSecret string `json:"databaseSecret"`
}

// Every components inherited from  HarborDeployment
type HarborDeployment struct {
	// Number of desired pods. This is a pointer to distinguish between explicit
	// zero and not specified. Defaults to 1.
	// +optional
	// +kubebuilder:validation:Minimum=1
	Replicas *int32 `json:"replicas,omitempty"`

	// +optional
	Image *string `json:"image,omitempty"`

	// +optional
	NodeSelector     NodeSelector                  `json:"nodeSelector,omitempty"`
	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
}
```

So, can we use `HarborDeployment` loop to set the common properties such as `Replicas `

```go
func resetReplicas(spec *HarborSpec) {
  for _, comp := range spec.Componnets {
      if hComp, ok := comp.(HarborDeployment); ok {
             hComp.Replicas = nil
     }
  }
}

func setReplica(spec *HarborSpec, expectedR int) {
   // similar logic with above reset func
}
```

Comment (wangcanfengxs) [reply controllers/harbor/harbor.go:61]: It need to transform HarborComponents to []HarborDeployment

File: controllers/harbor/harbor.go (hunk 1 above)
Comment (soulseen) [reply controllers/harbor/harbor.go]: It can't be handled with `switch` function in one reconciler process when it contains both updates and scale events.

**Pull request merged**